---
title: "TODO"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```




```{r setup}
options(warn = 1)
library(restrictedROC)
```
```{r, prepare data}
outdir <- "intermediate_data/2022-zhang_multivariate"
# data("dataMelanoma")
# This feature selection follows exactly "data-raw/2022-Zhang-GenomeMedicine/data/GM/Code/Model_Stem.Sig.R"
# Note that NOT all of the features from zhang2022_Stem.Sig are in the dataset!
train_stemsig <- dataMelanoma::zhang2022_training[
    ,
    colnames(dataMelanoma::zhang2022_training) %in% c("response", dataMelanoma::zhang2022_Stem.Sig)
]
val_stemsig <- dataMelanoma::zhang2022_validation[
    ,
    colnames(dataMelanoma::zhang2022_validation) %in% c("response", dataMelanoma::zhang2022_Stem.Sig)
]
test_stemsig <- dataMelanoma::zhang2022_test[
    ,
    colnames(dataMelanoma::zhang2022_test) %in% c("response", dataMelanoma::zhang2022_Stem.Sig)
]
data_tvt <- list("train" = train_stemsig, "val" = val_stemsig, "test" = test_stemsig)
rf_seed <- 4321
```
```{r, Usual randomforest}
baseline_rf <- restrictedROC::train_rROC_h2o(
    x_prepared = train_stemsig[, -1],
    y = train_stemsig$response,
    h2o_trainfun = function(df, col_y, cols_x, ...) {
        h2o::h2o.randomForest(
            training_frame = df,
            y = col_y,
            x = cols_x,
            ntrees = 1000,
            max_depth = 20,
            min_rows = 1,
            nbins = 20,
            seed = rf_seed, ...
        )
    }
)
rf_base_applied <- lapply(data_tvt, function(data_x) {
    restrictedROC::predict_rROC_h2o(
        h2o_model = baseline_rf,
        x_prepared = data_x[, -1],
        y = data_x$response
    )
})
```



```{r, Restriction preprocessing, then randomforest}
removed_impute <- -20
rroc_res <- restrictedROC::rROC(
    x = train_stemsig[, -1],
    y = train_stemsig[[1]],
    independent_vars = NULL,
    positive_label = "R",
    n_permutations = 0
)
prepared_df <- restrictedROC::prepare_modeldata(
    x = train_stemsig[, -1],
    y = train_stemsig[[1]],
    which_preds = c("bounded"),
    positive_label = "R",
    rroc_result = rroc_res,
    # range(c(test_stemsig[, -1], val_stemsig[, -1], train_stemsig[, -1]))
    # -18.71309  27.50762
    removed_impute = removed_impute
)
rf_postRestriction <- restrictedROC::train_rROC_h2o(
    x_prepared = prepared_df,
    y = train_stemsig[[1]],
    h2o_trainfun = function(df, col_y, cols_x, ...) {
        h2o::h2o.randomForest(
            training_frame = df,
            y = col_y,
            x = cols_x,
            ntrees = 1000,
            max_depth = 20,
            min_rows = 1,
            nbins = 20,
            seed = rf_seed, ...
        )
    }
)

rf_postRestriction_applied <- lapply(data_tvt, function(data_x) {
    prepared_data <- restrictedROC::prepare_modeldata(
        x = data_x[, -1],
        y = data_x[[1]],
        which_preds = c("bounded"),
        positive_label = "R",
        rroc_result = rroc_res,
        removed_impute = removed_impute
    )
    restrictedROC::predict_rROC_h2o(
        h2o_model = rf_postRestriction,
        x_prepared = prepared_data,
        y = data_x[[1]],
    )
})
```


```{r, Compare results}
reslist <- list(
    "base" = rf_base_applied,
    "restricted" = rf_postRestriction_applied
)
lapply(reslist, function(x) {
    lapply(x, function(y) {
        y[["metrics"]][["all"]]@metrics[["cm"]][["table"]]
    })
})
sapply(reslist, function(x) {
    sapply(x, function(y) {
        y[["metrics"]][["all"]]@metrics[["AUC"]]
    })
})

pdf(file.path(outdir, "rocs_default_multivariate.pdf"), height = 3.6, width = 3)
for (tvt_x in names(reslist[["base"]])) {
    preds_base <- reslist[["base"]][[tvt_x]][["predictions"]]
    preds_restricted <- reslist[["restricted"]][[tvt_x]][["predictions"]]
    roc_base <- pROC::roc(response = preds_base$y, predictor = preds_base[["R"]])
    roc_restricted <- pROC::roc(response = preds_restricted$y, predictor = preds_restricted[["R"]])
    cat("\n\n", tvt_x, "\n")
    compare_rocs <- pROC::roc.test(roc_base, roc_restricted, method = "delong", alternative = "less")
    print(compare_rocs)
    print(
        pROC::ggroc(roc_base, legacy.axes = TRUE) +
            ggpubr::theme_pubr() +
            ggplot2::geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
            ggplot2::ggtitle(
                paste0("Multivariate default params; ", tvt_x, " base"),
                subtitle = paste0(
                    "pval ROC base < restricted: ", round(compare_rocs$p.value, 4),
                    " --- AUC: ", round(roc_base$auc, 4)
                )
            ) +
            ggplot2::theme(
                plot.title = ggplot2::element_text(size = 5),
                plot.subtitle = ggplot2::element_text(size = 3)
            )
    )
    print(
        pROC::ggroc(roc_restricted, legacy.axes = TRUE) +
            ggpubr::theme_pubr() +
            ggplot2::geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
            ggplot2::ggtitle(
                paste0("Multivariate default params; ", tvt_x, " restricted"),
                subtitle = paste0(
                    "pval ROC base < restricted: ", round(compare_rocs$p.value, 4),
                    " --- AUC: ", round(roc_restricted$auc, 4)
                )
            ) +
            ggplot2::theme(
                plot.title = ggplot2::element_text(size = 5),
                plot.subtitle = ggplot2::element_text(size = 3)
            )
    )
    print(
        
        pROC::ggroc(list("base" = roc_base, "restricted"= roc_restricted), legacy.axes = TRUE) +
            ggpubr::theme_pubr() +
            ggplot2::geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
            ggplot2::ggtitle(
                paste0("Multivariate default params; ", tvt_x, " restricted"),
                subtitle = paste0(
                    "pval ROC base < restricted: ", round(compare_rocs$p.value, 4),
                    " --- AUC.base: ", round(roc_base$auc, 4),
                    " --- AUC.restricted: ", round(roc_restricted$auc, 4)
                )
            ) +
            ggplot2::theme(
                plot.title = ggplot2::element_text(size = 5),
                plot.subtitle = ggplot2::element_text(size = 3)
            )
    )
}
dev.off()




res_tables <- lapply(reslist[c("base", "restricted")], function(x) {
    lapply(x, function(y) {
        tmp <- y[["predictions"]][, -c(1:369)][, c("y", "predict")]
        tmp[["y"]] <- relevel(tmp[["y"]], ref = "R")
        tmp[["predict"]] <- relevel(tmp[["predict"]], ref = "R")
        return(
            list(
                table(tmp),
                "acc" = sum(diag(table(tmp))) / nrow(tmp) * 1.
            )
        )
    })
})
# Accuracies when taking the predictions of the random forest model.
# The class is denoted by the bigger probability (= cutoff is 0.5)
sapply(res_tables, function(x) sapply(x, function(y) y[["acc"]]))
#            base restricted
# train 1.0000000  1.0000000
# val   0.5064935  0.6363636
# test  0.5100671  0.5637584
```
[1]: Zhang, Z., Wang, ZX., Chen, YX. et al. Integrated analysis of single-cell and bulk RNA sequencing data reveals a pan-cancer stemness signature predicting immunotherapy response. Genome Med 14, 45 (2022). https://doi.org/10.1186/s13073-022-01050-w 
