---
title: "Datatypes 01: Proteomics"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

In <cite>[Harel et al, 2019, Cell][1]</cite>, the authors describe melanoma response to immunotherapy. They use high-resolution mass spectrometry to quantify 4,300 proteins across 116 patients. Formalin-fixed paraffin-embedded (FFPE) tissues from 42 metastatic melanoma patients treated with TIL-based adoptive cell transfer and 74 treated with anti-PD1 immunotherapy were included. 
Samples were taken before treatment except for two patients taken 11 or 18 days after treatment initialization. Patients are split into 21/40 responders, 21/27 non-responders and 0/7 stable disease according to RECIST v1.0 guidelines (TIL/anti-PD1 cohorts). See ``Tumor sample collection`` in the paper's method section. 

We used the data available under [Table S1](https://www.sciencedirect.com/science/article/pii/S0092867419309006?via%3Dihub#app2) of the publication in ``1-s2.0-S0092867419309006-mmc1.xlsx``, sheet ``S1C`` and ``S1D``.
According to the section ``Proteomics statistical analysis`` in the paper's method section, the following analysis was performed . 

    For the clinical tumor data, normalized ratio tumor/SILAC data were log2-transformed and the protein groups were filtered to have at least 70% valid values, reaching a list of 4588,4620 or 4416 protein groups for the TIL, the anti-PD1 or the combined datasets respectively, which were further used for all downstream analyses. Dataset integration was based on gene name; multiple entries for the same gene name were integrated to a single entry by calculating the median expression value. Data were normalized by subtracting most frequent value in each sample. To extract DEPs between responders and non-responders multiple comparisons were performed; a two-sample Student's t test was performed with a p value threshold of 0.05 for each of the two datasets, following by 2D annotation enrichment test (Cox and Mann, 2012) with FDR q-value 0.02 to derive the differential functional groups in each of the two datasets. A two-sample Student's t test with a permutation-based FDR q-value of 0.1 and S0 (Tusher et al., 2001) of 0.1 was performed for the anti-PD1 dataset or the combined dataset of both treatments, to extract significantly changing proteins. 

From main text; ``Functional Analysis of Responders and Non-responders to Immunotherapy``:

    Initial bioinformatic analysis examined the functional differences between responders and non-responders in each therapy. We first applied a low stringency Student's t test with a nominal p value cutoff (p value < 0.05) and found 414 and 636 differentially expressed proteins (DEPs) in the TIL and the anti-PD1 cohorts [...].



```{r setup}
options(warn = 1)
library(restrictedROC)
```

```{r download data}
library(dplyr)
dir.create("intermediate_data/2019-harel", recursive = TRUE)
harel_pheno <- dataMelanoma::harel2019_pheno
outcome_noSD <- dataMelanoma::harel2019_responder_noSD
```

```{r, Read S1C(TIL) and S1D(aPD1), filtered for 70% valid values and gene names median-merged}
data_mats <- dataMelanoma::harel2019_protmats

lapply(data_mats, colnames)
lapply(data_mats, dim)
# $aPD1
# [1] 4620   67
# $til
# [1] 4588   42
```

```{r, Reproduce t-tests}
harel_ttests <- lapply(data_mats, function(datapart_x) {
    res_t <- apply(datapart_x, 1, function(feature_x) {
        # names(x) are the column names of the matrix
        # The response needs to be ordered according to the column names
        x_NonResponder <- split(feature_x, outcome_noSD[names(feature_x)])
        res_test <- t.test(
            x_NonResponder[["responder"]],
            x_NonResponder[["non-responder"]],
            alternative = "two.sided",
            paired = FALSE,
            var.equal = TRUE
        )
        return(c(
            "neglog_p.value" = -log10(res_test$p.value),
            "p.value" = res_test$p.value,
            "t" = res_test$statistic[[1]],
            "diff_nr_r" = (res_test$estimate["mean of x"] - res_test$estimate["mean of y"])[[1]]
        ))
    })
    return(tibble::as_tibble(t(res_t), rownames = "gene") |> dplyr::arrange(diff_nr_r))
})
harel_ttests
lapply(harel_ttests, function(x) sum(x$p.value <= 0.05, na.rm = TRUE))
sign_ttests <- lapply(harel_ttests, function(x) x[x$p.value <= 0.05, na.rm = TRUE])

significant_genes_paper <- sapply(c("S2A", "S2B"), function(x) {
    tmp <- readxl::read_excel("intermediate_data/harel2019-s2.xlsx", sheet = x, skip = 1, na = "NaN")
    return(tmp[, grepl("Gene name", colnames(tmp))][[1]])
}, USE.NAMES = TRUE, simplify = FALSE)

if (length(significant_genes_paper[[1]][!significant_genes_paper[[1]] %in% sign_ttests[["til"]][["gene"]]]) > 0) {
    stop("Some TIL genes have not been found")
}
if (length(significant_genes_paper[[2]][!significant_genes_paper[[2]] %in% sign_ttests[["aPD1"]][["gene"]]]) > 0) {
    stop("Some TIL genes have not been found")
}
# Calculate q-values (BH-method)
harel_ttests_q <- lapply(harel_ttests, function(x) {
    x[["q.value"]] <- p.adjust(x[["p.value"]], method = "BH")
    x[["neglog_q.value"]] <- -log10(x[["q.value"]])
    return(x)
})
```

```{r, Figure 2c: p-values and q-values}
library(ggplot2)
pdf("intermediate_data/2019-harel/fig3c_redone.pdf")
set.seed(120398)
print(
    lapply(names(harel_ttests), function(part_x) {
        ggplot(harel_ttests[[part_x]], aes(x = diff_nr_r, y = neglog_p.value)) +
            geom_point() +
            ggrepel::geom_text_repel(
                data = subset(
                    harel_ttests_q[[part_x]],
                    p.value < 0.01 | abs(diff_nr_r) > 1
                ),
                aes(label = gene),
                size = 1.5, col = "darkgreen",
                min.segment.length = 0,
                max.overlaps = 5
            ) +
            ggtitle(part_x) +
            geom_hline(yintercept = -log10(0.05), col = "red", linetype = "dashed")
    })
)
dev.off()


pdf("intermediate_data/2019-harel/fig3c-02-qvalues.pdf")
set.seed(120398)
print(
    lapply(names(harel_ttests_q), function(part_x) {
        ggplot(harel_ttests_q[[part_x]], aes(x = diff_nr_r, y = neglog_q.value)) +
            geom_point() +
            ggrepel::geom_text_repel(
                data = subset(
                    harel_ttests_q[[part_x]],
                    q.value < 0.1 | abs(diff_nr_r) > 1
                ),
                aes(label = gene),
                size = 1.5, col = "darkgreen",
                min.segment.length = 0,
                max.overlaps = 5
            ) +
            ggtitle(part_x) +
            geom_hline(yintercept = -log10(0.05), col = "red", linetype = "dashed")
    })
)
dev.off()
```

```{r, Restriction}
n_permutations <- 50000
int_dir <- "intermediate_data/2019-harel/rroc_intermediate"
dir.create(int_dir, recursive = TRUE)
library(future)
plan(multisession)
rroc_res <- sapply(
    names(data_mats),
    function(x_name) {
        x <- data_mats[[x_name]]
        x_noNArownames <- x[!is.na(rownames(x)), ]
        restrictedROC::rROC(
            x = t(x_noNArownames),
            y = list("responder" = outcome_noSD[colnames(x)]),
            n_permutations = n_permutations,
            positive_label = "responder",
            save_path = file.path(int_dir, paste0("harel_nPERM", n_permutations, "_", x_name, ".qs")),
            parallel_permutations = TRUE,
            verbose = TRUE
        )
    },
    USE.NAMES = TRUE, simplify = FALSE
)

# 50k permutations took 70s for one feature.
# There are 4620 (aPD1) and 4588 (TIL) features
# Therefore that takes 175h.
rroc_res_tmp <- sapply(names(data_mats), function(x_name) {
    qs::qread(file.path(int_dir, paste0("harel_nPERM", n_permutations, "_", x_name, ".qs")))
}, USE.NAMES = TRUE, simplify = FALSE)

rroc_res <- lapply(rroc_res_tmp, function(x) {
    summarized <- restrictedROC:::summary.rROC(x)
    summarized[["qval.twoside.max"]] <- p.adjust(summarized[["pval.twoside.max"]], method = "BH")
    summarized[["qval.twoside.global"]] <- p.adjust(summarized[["pval.twoside.global"]], method = "BH")
    summarized
})
```


```{r, Plotting global vs restricted}
library(ggplot2)
pval_cutoff <- .1
pdf("intermediate_data/2019-harel/global_max_pval.pdf", width = 4, height = 4)
for (name_x in names(rroc_res)) {
    p1 <- ggplot(rroc_res[[name_x]], aes(x = pval.twoside.global, y = pval.twoside.max))
    p2 <- ggplot(rroc_res[[name_x]], aes(x = qval.twoside.global, y = qval.twoside.max))
    for (plot_x in list(p1, p2)) {
        print(
            plot_x +
                geom_point() +
                geom_abline(slope = 1, intercept = 0, linewidth = .1) +
                geom_vline(xintercept = pval_cutoff, linetype = "dashed") +
                geom_hline(yintercept = pval_cutoff, linetype = "dashed") +
                scale_x_log10() +
                scale_y_log10() +
                # label feature with q.value < pval_cutoff
                ggrepel::geom_text_repel(
                    data = subset(
                        rroc_res[[name_x]],
                        qval.twoside.global > pval_cutoff & qval.twoside.max < pval_cutoff
                    ),
                    aes(label = level_2),
                    size = 1.5, col = "darkgreen",
                    min.segment.length = 0,
                    max.overlaps = 5
                ) +
                ggtitle(name_x) +
                ggpubr::theme_pubr()
        )
    }
}
dev.off()

source(file.path("vignettes", "articles", "paper_utils.R"))
library(dplyr)
# Plotting the restricted ROC
pdf("intermediate_data/2019-harel/rroc_significant.pdf", height = 6.8, width = 6.5)
sink("intermediate_data/2019-harel/rroc_significant_booktabs.tex")
for (feature_outcome in list(
    c("WDR47", "aPD1"),
    c("RQCD1", "aPD1"),
    c("ETFB", "til"),
    c("CALCOCO2", "til")
)) {
    feature_x <- feature_outcome[[1]]
    outcome_x <- feature_outcome[[2]]

    data_x <- data_mats[[outcome_x]][feature_x, ]
    data_x_noNA <- na.omit(data_x)
    grouped_vals <- split(data_x_noNA, outcome_noSD[names(data_x_noNA)])
    names(grouped_vals) <- gsub("-", "", names(grouped_vals))

    tmp <- paper_performance(
        values_grouped = grouped_vals,
        positive_label = "responder",
        plot_n_points = 0,
        current_rroc = rroc_res[[outcome_x]] |>
            dplyr::filter(level_2 == feature_x),
        outcome = outcome_x,
        feature = feature_x
    )
    p1 <- tmp[["plot"]]
    p1 <- p1 + patchwork::plot_annotation(
        title = paste0("Dataset: ", outcome_x, "  --- Feature: ", feature_x)
    )
    print(p1)
    cat(paperperf_to_tex(tmp[["perf"]]))
}
dev.off()
sink()
```

[1]: Harel et al. "Proteomics of melanoma response to immunotherapy reveals mitochondrial dependence." Cell 179.1 (2019): 236-250. https://doi.org/10.1016/j.cell.2019.08.012
