---
title: "Datatypes 04: Microbiome"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```




```{r setup}
options(warn = 1)
library(restrictedROC)
```
```{r data}
# sudo apt install libgsl-dev   # DirichletMultinomial needs this
pacman::p_install("DirichletMultinomial", force = FALSE)
pacman::p_install("curatedMetagenomicData", force = FALSE)

pacman::p_load(dplyr)
curatedMetagenomicData::sampleMetadata |>
    dplyr::filter(grepl("melanoma", disease, ignore.case = TRUE)) |>
    pull(study_name) |>
    unique()
# "FrankelAE_2017"       "GopalakrishnanV_2018" "LeeKA_2022"
# "MatsonV_2018"         "PetersBA_2019"        "WindTT_2020"

melanoma_studies_lee <-
    tibble::as_tibble(sampleMetadata) |>
    dplyr::filter(grepl("melanoma", disease, ignore.case = TRUE)) |>
    # We excluded any samples taken
    # [EXCLUDED:] after the start of ICI therapy
    # Not in metadata
    # [EXCLUDED:] nonmetagenomic samples
    # ??
    # [EXCLUDED:] nonfecal samples
    dplyr::filter(body_site == "stool") |>
    # [EXCLUDED:] samples with low sequencing depth(<1 million reads)
    dplyr::filter(number_reads > 1e6)

data_lee <- melanoma_studies_lee |> curatedMetagenomicData::returnSamples(dataType = "relative_abundance")
# colnames(melanoma_studies_lee)
# melanoma_studies_lee$days_from_first_collection
# melanoma_studies_lee$visit_number
# melanoma_studies_lee |> writexl::write_xlsx("melanoma_studies_lee.xlsx")

SummarizedExperiment::colData(data_lee) |>
    tibble::as_tibble() |>
    select(study_name, location) |>
    table(useNA = "always")
SummarizedExperiment::colData(data_lee) |>
    tibble::as_tibble() |>
    select(study_name, subcohort) |>
    table(useNA = "always")

sample_subcohort_info <- SummarizedExperiment::colData(data_lee) |>
    tibble::as_tibble() |>
    select(study_name, subcohort, location)
sample_subcohort_info[["lee_subcohort"]] <- NA_character_
for (sample_i in 1:nrow(sample_subcohort_info)) {
    current <- sample_subcohort_info[sample_i, ]
    if ((is.na(current["subcohort"]) && is.na(current["location"])) || current["study_name"] == "PetersBA_2019") {
        sample_subcohort_info[sample_i, ][["lee_subcohort"]] <- current[["study_name"]]
    } else if (current["study_name"] == "LeeKA_2022") {
        if (!is.na(current["location"])) {
            sample_subcohort_info[sample_i, ][["lee_subcohort"]] <- current[["location"]]
        } else {
            sample_subcohort_info[sample_i, ][["lee_subcohort"]] <- current[["subcohort"]]
        }
    } else {
        print(current)
        stop("I do not know the current sample's subcohort.")
    }
}
data_lee[["lee_subcohort"]] <- sample_subcohort_info[["lee_subcohort"]]

print(table(
    SummarizedExperiment::colData(data_lee)[["lee_subcohort"]],
    factor(SummarizedExperiment::colData(data_lee)[["ORR"]], levels = c("yes", "no")),
    useNA = "always"
))
print(table(
    SummarizedExperiment::colData(data_lee)[["lee_subcohort"]],
    factor(SummarizedExperiment::colData(data_lee)[["PFS12"]], levels = c("yes", "no")),
    useNA = "always"
))
print(apply(table(
    SummarizedExperiment::colData(data_lee)[["lee_subcohort"]],
    factor(SummarizedExperiment::colData(data_lee)[["PFS12"]], levels = c("yes", "no")),
    useNA = "always"
), 2, sum))
SummarizedExperiment::colData(data_lee) |>
    tibble::as_tibble() |>
    select(study_name, subcohort, location) |>
    print(n = 1000)

# data_lee[["subcohort_lee"]] <- SummarizedExperiment::colData(data_lee)[["subcohort"]]
# data_relab <- SummarizedExperiment::assay(data_lee) |>
#     tibble::as_tibble() |>
#     tibble::rownames_to_column()
# data_relab[["rowname"]] <- SummarizedExperiment::rownames(data_lee)



# pheno_orr <- melanoma_studies_lee |>
#     dplyr::filter(!is.na(ORR))

# pheno_orr |>
#     select(subcohort, ORR) |>
#     table(useNA = "always")
# pheno_noORR <- melanoma_studies_lee |>
#     dplyr::filter(is.na(ORR))
# pheno_noORR
```
```{r, helper functions}
#' @param x A matrix of features, rows = features, cols=samples
#'
#' @param response A vector of responses, length = samples
#' @param save_intermediate_file A file to save intermediate results to
#' @param ...
#' Further parameters to pass to \code{\link{restrictedROC::simple_rROC_permutation}}
#'
rroc_permutation_many_features <- function(x,
                                           response,
                                           save_intermediate_file = NA,
                                           verbose = FALSE,
                                           ...) {
    if (ncol(x) != length(response)) {
        stop("Number of columns in x must equal length of response")
    }

    tmp_reslist <- list()
    for (feature in rownames(x)) {
        cat("Feature: ", feature, "\n")
        tmp_reslist[[feature]] <- restrictedROC::simple_rROC_permutation(
            response = response,
            predictor = x[feature, ],
            ...
        )
    }
    tmp_reslist_formatted <- tibble::as_tibble(do.call(rbind, lapply(
        tmp_reslist, function(x) {
            return(c(
                "max_neglog_p.value" = -log10(x$permutation_pval[["pval.twoside.max"]]),
                "global_neglog_p.value" = -log10(x$permutation_pval[["pval.twoside.global"]]),
                "max_p.value" = (x$permutation_pval[["pval.twoside.max"]]),
                "global_p.value" = (x$permutation_pval[["pval.twoside.global"]]),
                "avg.FC" = x[["avg.FC"]]
            ))
        }
    )))
    tmp_reslist_formatted[["max_q.value"]] <- p.adjust(tmp_reslist_formatted[["max_p.value"]], method = "BH")
    tmp_reslist_formatted[["global_q.value"]] <- p.adjust(tmp_reslist_formatted[["global_p.value"]], method = "BH")
    tmp_reslist_formatted[["feature"]] <- names(tmp_reslist)
    if (!is.na(save_intermediate_file)) {
        qs::qsave(tmp_reslist_formatted, file = save_intermediate_file)
        cat("\nSaved ", save_intermediate_file, "\n")
    }
}
```

```{r, Fig2 - PFS12 univariate analysis (Fig 3d)}
# qs::qsavem(data_lee, sample_subcohort_info, melanoma_studies_lee, rroc_permutation_many_features, file = "intermediate_data/2022-lee/dev-data_lee.rds")
outdir <- "intermediate_data/2022-lee"
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)

data_pfs <- data_lee[, !is.na(SummarizedExperiment::colData(data_lee)[["PFS12"]])]
table_lsubcohort_pfs <- table(
    SummarizedExperiment::colData(data_lee)[["lee_subcohort"]],
    factor(SummarizedExperiment::colData(data_lee)[["PFS12"]], levels = c("yes", "no"))
)
library(future)
plan(multisession)
for (n_permutations in c(1e3, 25e3, 1e5)) {
    for (subcohort_XX in c("PRIMM-NL", "PRIMM-UK")) {
        cat("\n\nn_permutations: ", n_permutations, "- subcohort: ", subcohort_XX, "\n")
        samples_boolean <- SummarizedExperiment::colData(data_pfs)[["lee_subcohort"]] == subcohort_XX
        pheno_pfs_part <- SummarizedExperiment::colData(data_pfs)[samples_boolean, ]
        data_pfs_part <- data_pfs[, samples_boolean]
        rroc_permutation_many_features(
            x = SummarizedExperiment::assay(data_pfs_part)[1:3, ],
            response = pheno_pfs_part[["PFS12"]],
            save_intermediate_file = file.path(outdir, paste0(
                "lee2022_pfs_rroc_nperm.", n_permutations,
                "_SUB.", subcohort_XX, ".qrds"
            )),
            n_permutations = n_permutations,
            positive_label = "yes",
            parallel_permutations = TRUE,
            verbose = TRUE
        )
    }
}

read_lee <- qs::qread("intermediate_data/lee2022_pfs_rroc_1e+05.rds")
```

```{r leave-one-out loop}
wrapper <- function(
    summarized_experiment,
    boolean_train,
    boolean_test) {
    if (length(boolean_train) != length(boolean_test)) {
        stop("boolean_train and boolean_test must be the same length")
    }
    if (length(booleant_train) != ncol(summarized_experiment)) {
        stop("boolean_train must be the same length as the number of columns in summarized_experiment")
    }
}

for (data_x in list(data_pfs)) {
    for (subcohort_XX in rownames(table_lsubcohort_pfs)) {
        cat("\n\nsubcohort: ", subcohort_XX, "\n")
        samples_boolean_test <- SummarizedExperiment::colData(data_x)[["lee_subcohort"]] != subcohort_XX
        samples_boolean_train <- SummarizedExperiment::colData(data_x)[["lee_subcohort"]] == subcohort_XX

        stop()
        pheno_pfs_part <- SummarizedExperiment::colData(data_x)[samples_boolean, ]
        data_x_part <- data_x[, samples_boolean]
        rroc_permutation_many_features(
            x = SummarizedExperiment::assay(data_x_part)[1:3, ],
            response = pheno_pfs_part[["PFS12"]],
            save_intermediate_file = file.path(outdir, paste0(
                "dev-lee2022_pfs_rroc_nperm.", n_permutations,
                "_SUB.", subcohort_XX, ".qrds"
            )),
            n_permutations = n_permutations,
            positive_label = "yes",
            parallel_permutations = TRUE,
            verbose = TRUE
        )
    }
}
```

[1]: Lee, K.A., Thomas, A.M., Bolte, L.A. et al. Cross-cohort gut microbiome associations with immune checkpoint inhibitor response in advanced melanoma. Nat Med 28, 535â€“544 (2022). https://doi.org/10.1038/s41591-022-01695-5
