---
title: "Datatypes 01: Proteomics"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

In <cite>[Harel et al, 2019, Cell][1]</cite>, the authors describe melanoma response to immunotherapy. They use high-resolution mass spectrometry to quantify 4,300 proteins across 116 patients. Formalin-fixed paraffin-embedded (FFPE) tissues from 42 metastatic melanoma patients treated with TIL-based adoptive cell transfer and 74 treated with anti-PD1 immunotherapy were included. 
Samples were taken before treatment except for two patients taken 11 or 18 days after treatment initialization. Patients are split into 21/40 responders, 21/27 non-responders and 0/7 stable disease according to RECIST v1.0 guidelines (TIL/anti-PD1 cohorts). See ``Tumor sample collection`` in the paper's method section. 

We used the data available under [Table S2](https://www.sciencedirect.com/science/article/pii/S0092867419309006?via%3Dihub#app2) of the publication in ``1-s2.0-S0092867419309006-mmc1.xlsx``, sheet ``S1C`` and ``S1D``.
According to the section ``Proteomics statistical analysis`` in the paper's method section, the following analysis was performed . 

    For the clinical tumor data, normalized ratio tumor/SILAC data were log2-transformed and the protein groups were filtered to have at least 70% valid values, reaching a list of 4588,4620 or 4416 protein groups for the TIL, the anti-PD1 or the combined datasets respectively, which were further used for all downstream analyses. Dataset integration was based on gene name; multiple entries for the same gene name were integrated to a single entry by calculating the median expression value. Data were normalized by subtracting most frequent value in each sample. To extract DEPs between responders and non-responders multiple comparisons were performed; a two-sample Student's t test was performed with a p value threshold of 0.05 for each of the two datasets, following by 2D annotation enrichment test (Cox and Mann, 2012) with FDR q-value 0.02 to derive the differential functional groups in each of the two datasets. A two-sample Student's t test with a permutation-based FDR q-value of 0.1 and S0 (Tusher et al., 2001) of 0.1 was performed for the anti-PD1 dataset or the combined dataset of both treatments, to extract significantly changing proteins. 

From main text; ``Functional Analysis of Responders and Non-responders to Immunotherapy``:

    Initial bioinformatic analysis examined the functional differences between responders and non-responders in each therapy. We first applied a low stringency Student's t test with a nominal p value cutoff (p value < 0.05) and found 414 and 636 differentially expressed proteins (DEPs) in the TIL and the anti-PD1 cohorts [...].



```{r setup}
options(warn = 1)
library(restrictedROC)
```

```{r download data}
dir.create("intermediate_data")
download.file(
    url = "https://ars.els-cdn.com/content/image/1-s2.0-S0092867419309006-mmc1.xlsx",
    destfile = "intermediate_data/harel2019.xlsx",
)
# S1C and S1D are the features with at least 70% present values
harel_til <- readxl::read_excel("intermediate_data/harel2019.xlsx", sheet = "S1C", skip = 1, na = "NaN")
harel_aPD1 <- readxl::read_excel("intermediate_data/harel2019.xlsx", sheet = "S1D", skip = 1, na = "NaN")

# The description columns start with a T
harel_til_mat <- as.matrix(harel_til[, !grepl("^T:", colnames(harel_til))])
rownames(harel_til_mat) <- sub(";.*", "", harel_til[["T: Gene names"]])

harel_aPD1_mat <- as.matrix(harel_aPD1[, !grepl("^T:", colnames(harel_aPD1))])
rownames(harel_aPD1_mat) <- sub(";.*", "", harel_aPD1[["T: T: Gene name"]])
```

```{r Reproduce t-tests}
```

[1]: Harel et al. "Proteomics of melanoma response to immunotherapy reveals mitochondrial dependence." Cell 179.1 (2019): 236-250. https://doi.org/10.1016/j.cell.2019.08.012
